File: hw/carvajal/u2/hw19Inheritance/JSFarmSystem/model/FarmAnimal.js
export class FarmAnimal {
    constructor(id, breed, bornOn, gender, isAbleToReproduce, weight, cage) {
        this.id = id;
        this.breed = breed;
        this.bornOn = bornOn;
        this.gender = gender;
        this.isAbleToReproduce = isAbleToReproduce;
        this.weight = weight;
        this.cage = cage;
    }

    getAgeInMonths() {
        const now = new Date();
        const diff = (now - this.bornOn) / (1000 * 60 * 60 * 24 * 30);
        return Math.floor(diff);
    }

    assignCage(cage) {
        this.cage = cage;
    }

    toString() {
        return `FarmAnimal{id=${this.id}, breed='${this.breed}', bornOn=${this.bornOn}, gender=${this.gender}, isAbleToReproduce=${this.isAbleToReproduce}, weight=${this.weight}, cage=${this.cage}}`;
    }
}

File: hw/calle/u2/hw19Inheritance/FarmsystemJava/src/main/java/ec/edu/espe/farmsystem/model/FarmAnimal.java
package ec.edu.espe.farmsystem.model;

import java.time.Instant;
import java.time.LocalDate;
import java.time.Period;
import java.time.ZoneId;
import java.util.Date;

public abstract class FarmAnimal {
    private int id;
    private String breed;
    private Date bornOn;
    private String gender;
    private boolean isAbleToReproduce;
    private float weight;
    private Cage cage;
    
    public FarmAnimal(int id, String breed, Date bornOn, String gender, boolean isAbleToReproduce, float weight, Cage cage) {
        this.id = id;
        this.breed = breed;
        this.bornOn = bornOn;
        this.gender = gender;
        this.isAbleToReproduce = isAbleToReproduce;
        this.weight = weight;
        this.cage = cage;
    }
    
    public int getAgeInMonths(){
        // Convierte java.util.Date a LocalDate para calcular la edad con java.time
        Instant instant = bornOn.toInstant();
        LocalDate bornDate = instant.atZone(ZoneId.systemDefault()).toLocalDate();
        LocalDate currentDate = LocalDate.now();
        
        Period period = Period.between(bornDate, currentDate);
        
        int totalMonths = period.getYears() * 12 + period.getMonths();
        
        return totalMonths;
    }
    
    public void assignCage(Cage cage){
        this.setCage(cage);
    }

    @Override
    public String toString() {
        return "FarmAnimal{id=" + id + ", breed=" + breed + ", bornOn=" + bornOn + ", gender=" + gender + ", isAbleToReproduce=" + isAbleToReproduce + ", weight=" + weight + ", cage=" + cage + ", ageInMonths=" + getAgeInMonths() + '}';
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getBreed() {
        return breed;
    }

    public void setBreed(String breed) {
        this.breed = breed;
    }

    public Date getBornOn() {
        return bornOn;
    }

    public void setBornOn(Date bornOn) {
        this.bornOn = bornOn;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public boolean isIsAbleToReproduce() {
        return isAbleToReproduce;
    }

    public void setIsAbleToReproduce(boolean isAbleToReproduce) {
        this.isAbleToReproduce = isAbleToReproduce;
    }

    public float getWeight() {
        return weight;
    }

    public void setWeight(float weight) {
        this.weight = weight;
    }

    public Cage getCage() {
        return cage;
    }

    public void setCage(Cage cage) {
        this.cage = cage;
    }
    
    
}